subsys/bluetooth/services/ams_client.c

```diff
@@ -0,0 +1,551 @@
+					 const void *data, uint16_t length)
+{
+	struct bt_ams_client *ams_c;
+	const uint8_t *value = (const uint8_t *)data;
+	struct bt_ams_entity_update_notif notif;
+	int err;
+
+	/* Retrieve AMS client module context. */
+	ams_c = CONTAINER_OF(params, struct bt_ams_client, notif_params_eu);
+
+	err = (length >= 3) ? 0 : -EINVAL;
+	if (!err) {
+		notif.ent_attr.entity = (enum bt_ams_entity_id)*(value + 0);
+		switch (notif.ent_attr.entity) {
+		case BT_AMS_ENTITY_ID_PLAYER:
+			notif.ent_attr.attribute.player =
+				(enum bt_ams_player_attribute_id)*(value + 1);
+			break;
+		case BT_AMS_ENTITY_ID_QUEUE:
+			notif.ent_attr.attribute.queue =
+				(enum bt_ams_queue_attribute_id)*(value + 1);
+			break;
+		case BT_AMS_ENTITY_ID_TRACK:
+			notif.ent_attr.attribute.track =
+				(enum bt_ams_track_attribute_id)*(value + 1);
+			break;
+		default:
+			err = -EINVAL;
+			break;
+		}
+		notif.flags = *(value + 2);
+		notif.data = value + 3;
+		notif.len = length - 3;
+	}
+
+	if (ams_c->notify_eu_cb) {
+		ams_c->notify_eu_cb(ams_c, &notif, err);
+	}
+
+	return BT_GATT_ITER_CONTINUE;
+}
+
+int bt_ams_subscribe_entity_update(struct bt_ams_client *ams_c,
+				   bt_ams_entity_update_notify_cb func)
+{
+	int err;
+
+	if (!ams_c || !func) {
+		return -EINVAL;
+	}
+	if (!ams_c->conn) {
+		return -EINVAL;
+	}
+
+	if (atomic_test_and_set_bit(&ams_c->state, AMS_EU_NOTIF_ENABLED)) {
+		return -EALREADY;
+	}
+
+	ams_c->notify_eu_cb = func;
+
+	ams_c->notif_params_eu.notify = bt_ams_eu_notify_callback;
+	ams_c->notif_params_eu.value = BT_GATT_CCC_NOTIFY;
+	ams_c->notif_params_eu.value_handle = ams_c->handle_eu;
+	ams_c->notif_params_eu.ccc_handle = ams_c->handle_eu_ccc;
+	atomic_set_bit(ams_c->notif_params_eu.flags,
+		       BT_GATT_SUBSCRIBE_FLAG_VOLATILE);
+
+	err = bt_gatt_subscribe(ams_c->conn, &ams_c->notif_params_eu);
+	if (err) {
+		atomic_clear_bit(&ams_c->state, AMS_EU_NOTIF_ENABLED);
+		LOG_ERR("Subscribe Entity Update failed (err %d)", err);
+	} else {
+		LOG_DBG("Entity Update subscribed");
+	}
+
+	return err;
+}
+
+int bt_ams_unsubscribe_entity_update(struct bt_ams_client *ams_c)
+{
+	int err;
+
+	if (!ams_c) {
+		return -EINVAL;
+	}
+
+	if (!atomic_test_bit(&ams_c->state, AMS_EU_NOTIF_ENABLED)) {
+		return -EFAULT;
+	}
+
+	err = bt_gatt_unsubscribe(ams_c->conn, &ams_c->notif_params_eu);
+	if (err) {
+		LOG_ERR("Unsubscribe Entity Update failed (err %d)", err);
+	} else {
+		atomic_clear_bit(&ams_c->state, AMS_EU_NOTIF_ENABLED);
+		LOG_DBG("Entity Update unsubscribed");
+	}
+
+	return err;
+}
+
+static void bt_ams_rc_write_callback(struct bt_conn *conn, uint8_t err,
+				      struct bt_gatt_write_params *params)
+{
+	struct bt_ams_client *ams_c;
+	bt_ams_write_cb write_callback;
+
+	/* Retrieve AMS client module context. */
+	ams_c = CONTAINER_OF(params, struct bt_ams_client, write_params_rc);
+
+	write_callback = ams_c->write_rc_cb;
+	atomic_clear_bit(&ams_c->state, AMS_RC_WRITE_PENDING);
+	if (write_callback) {
```

Comment:
After the bit AMS_RC_WRITE_PENDING is cleared, it is possible that another thread calls bt_ams_write_remote_command() and set the write_rc_cb variable to another value. The write_callback variable is saving a copy of write_rc_cb variable before it can be overwritten by another thread. It prevents a racing condition that may cause lengthy debugging later.